// vscode/vscode/src/vs/workbench/contrib/accessibility/browser/accessibilityConfiguration.js
import { localize } from 'monaco-editor/esm/vs/nls.js';
import { Extensions } from 'monaco-editor/esm/vs/platform/configuration/common/configurationRegistry.js';
import { Registry } from 'monaco-editor/esm/vs/platform/registry/common/platform.js';
import { RawContextKey } from 'monaco-editor/esm/vs/platform/contextkey/common/contextkey.js';
import { workbenchConfigurationNodeBase } from '../../../common/configuration.js';
const accessibilityHelpIsShown = ( new RawContextKey('accessibilityHelpIsShown', false, true));
const accessibleViewIsShown = ( new RawContextKey('accessibleViewIsShown', false, true));
const accessibleViewSupportsNavigation = ( new RawContextKey('accessibleViewSupportsNavigation', false, true));
const accessibleViewVerbosityEnabled = ( new RawContextKey('accessibleViewVerbosityEnabled', false, true));
const accessibleViewGoToSymbolSupported = ( new RawContextKey('accessibleViewGoToSymbolSupported', false, true));
const accessibleViewCurrentProviderId = ( new RawContextKey('accessibleViewCurrentProviderId', undefined, undefined));
const baseProperty = {
    type: 'boolean',
    default: true,
    tags: ['accessibility']
};
const configuration = {
    id: 'accessibility',
    title: ( localize('accessibilityConfigurationTitle', "Accessibility")),
    type: 'object',
    properties: {
        ["accessibility.verbosity.terminal" ]: {
            description: ( localize(
                'verbosity.terminal.description',
                'Provide information about how to access the terminal accessibility help menu when the terminal is focused'
            )),
            ...baseProperty
        },
        ["accessibility.verbosity.diffEditor" ]: {
            description: ( localize(
                'verbosity.diffEditor.description',
                'Provide information about how to navigate changes in the diff editor when it is focused'
            )),
            ...baseProperty
        },
        ["accessibility.verbosity.panelChat" ]: {
            description: ( localize(
                'verbosity.chat.description',
                'Provide information about how to access the chat help menu when the chat input is focused'
            )),
            ...baseProperty
        },
        ["accessibility.verbosity.inlineChat" ]: {
            description: ( localize(
                'verbosity.interactiveEditor.description',
                'Provide information about how to access the inline editor chat accessibility help menu and alert with hints which describe how to use the feature when the input is focused'
            )),
            ...baseProperty
        },
        ["accessibility.verbosity.inlineCompletions" ]: {
            description: ( localize(
                'verbosity.inlineCompletions.description',
                'Provide information about how to access the inline completions hover and accessible view'
            )),
            ...baseProperty
        },
        ["accessibility.verbosity.keybindingsEditor" ]: {
            description: ( localize(
                'verbosity.keybindingsEditor.description',
                'Provide information about how to change a keybinding in the keybindings editor when a row is focused'
            )),
            ...baseProperty
        },
        ["accessibility.verbosity.notebook" ]: {
            description: ( localize(
                'verbosity.notebook',
                'Provide information about how to focus the cell container or inner editor when a notebook cell is focused.'
            )),
            ...baseProperty
        },
        ["accessibility.verbosity.hover" ]: {
            description: ( localize(
                'verbosity.hover',
                'Provide information about how to open the hover in an accessible view.'
            )),
            ...baseProperty
        },
        ["accessibility.verbosity.notification" ]: {
            description: ( localize(
                'verbosity.notification',
                'Provide information about how to open the notification in an accessible view.'
            )),
            ...baseProperty
        },
        ["accessibility.verbosity.untitledHint" ]: {
            description: ( localize(
                'verbosity.untitledhint',
                'Provide information about relevant actions in an untitled text editor.'
            )),
            ...baseProperty
        }
    }
};
function registerAccessibilityConfiguration() {
    const registry = ( Registry.as(Extensions.Configuration));
    registry.registerConfiguration(configuration);
    registry.registerConfiguration({
        ...workbenchConfigurationNodeBase,
        properties: {
            ["accessibility.dimUnfocused.enabled" ]: {
                description: ( localize(
                    'dimUnfocusedEnabled',
                    'Whether to dim unfocused editors and terminals, which makes it more clear where typed input will go to. This works with the majority of editors with the notable exceptions of those that utilize iframes like notebooks and extension webview editors.'
                )),
                type: 'boolean',
                default: false,
                tags: ['accessibility'],
                scope: 1 ,
            },
            ["accessibility.dimUnfocused.opacity" ]: {
                markdownDescription: ( localize(
                    'dimUnfocusedOpacity',
                    'The opacity fraction (0.2 to 1.0) to use for unfocused editors and terminals. This will only take effect when {0} is enabled.',
                    `\`#${"accessibility.dimUnfocused.enabled" }#\``
                )),
                type: 'number',
                minimum: 0.2 ,
                maximum: 1 ,
                default: 0.75 ,
                tags: ['accessibility'],
                scope: 1 ,
            }
        }
    });
}
export { accessibilityHelpIsShown, accessibleViewCurrentProviderId, accessibleViewGoToSymbolSupported, accessibleViewIsShown, accessibleViewSupportsNavigation, accessibleViewVerbosityEnabled, registerAccessibilityConfiguration };
